_source/_posts/2022-07-21-angular-security-xss.md

```diff
@@ -0,0 +1,402 @@
+tags: [security, angular]
+tweets:
+- ""
+- ""
+- ""
+image:
+type: awareness
+---
+
+In the last post of this SPA security series, we covered Cross-Site Request Forgery (CSRF) and how Angular helps you with a mitigation technique. 
+
+|Posts in the SPA web security series|
+| --- |
+| 1. [Defend Your SPA from Security Woes](/blog/2022/07/06/spa-web-security) |
+| 2. [Defend Your SPA from Common Web Attacks](/blog/2022/07/08/spa-web-securty-csrf-xss) |
+| 3. [Protect Your Angular App from Cross-Site Request Forgery](/blog/2022/07/19/angular-security-csrf) |
+| 4. **Protect Your Angular App from Cross-Site Scripting** |
+
+Next, we'll dive into Cross-Site Scripting (XSS) and look at the built-in security guards you get when using Angular.
+
+{% include toc.md %}
+
+## Cross-Site Scripting (XSS) protection
+
+In the [second post of this series](/blog/2022/07/08/spa-web-securty-csrf-xss), we covered an overview of Cross-Site Scripting (XSS). In summary, you learned that XSS occurs when code pollutes data and your application doesn't take safeguards to prevent the code from running.
+
+Let's recap the example attack vector.
+
+> Imagine an overly dramatic but otherwise innocent scenario like this
+> 1. A website allows you to add comments about your favorite K-Drama.
+> 2. An agitator adds the comment `<script>alert('Crash Landing on You stinks!');</script>`.
+> 3. That terrible comment saves as is to the database.
+> 4. A K-Drama fan opens the website.
+> 5. The terrible comment is added to the website, appending the `<script></script>` tag to the DOM.
+> 6. The K-Drama fan is outraged by the JavaScript alert saying their favorite K-Drama, [Crash Landing on You](https://www.imdb.com/title/tt10850932/), stinks.
+
+In this example, we have a `<script>` element and glossed over the steps for appending the element to the DOM. In reality, there's a variety of ways that polluted data gets pulled into the application. The main culprits are adding untrusted data in [injection sinks](https://developer.mozilla.org/en-US/docs/Web/API/Trusted_Types_API#injection_sinks) - a Web API function that allows us to add dynamic content to our applications. Examples of sinks include, but aren't limited to:
+ * methods to append to the DOM such as  `innerHTML`, `outerHTML`
+ * approaches that load external resources or navigate to external sites via a URL, such as `src` or `href` for HTML elements and `url` property on styles
+ * event handlers, such as `onmouseover` and `onerror` with an invalid `src` value
+ * global functions that evaluate and/or run code, such as `eval()`, `setTimeout()`
+
+As you can see, there are many vectors for the vulnerability. Many of these sinks have legitimate use cases when building dynamic web applications. Since these sinks are necessary for web app functionality, we must use trusted data by escaping and sanitizing it.
```

Comment:
```suggestion
As you can see, there are many vectors for the vulnerability. Many of these sinks have legitimate use cases when building dynamic web applications. Since the sinks are necessary for web app functionality, we must use trusted data by escaping and sanitizing it.
```